Modules/_testcapimodule.c

```diff
@@ -1159,6 +1159,39 @@ test_get_type_name(PyObject *self, PyObject *Py_UNUSED(ignored))
 }
 
 
+static PyObject *
+test_get_type_qualname(PyObject *self, PyObject *Py_UNUSED(ignored))
+{
+    PyObject *tp_qualname = PyType_GetQualName(&PyLong_Type);
+    assert(strcmp(PyUnicode_AsUTF8(tp_qualname), "int") == 0);
+    Py_DECREF(tp_qualname);
+
+    tp_qualname = PyType_GetQualName(&_PyNamespace_Type);
+    assert(strcmp(PyUnicode_AsUTF8(tp_qualname), "SimpleNamespace") == 0);
+    Py_DECREF(tp_qualname);
+
+    PyObject *HeapTypeNameType = PyType_FromSpec(&HeapTypeNameType_Spec);
+    if (HeapTypeNameType == NULL) {
+        Py_RETURN_NONE;
+    }
+    PyObject *spec_name = PyUnicode_FromString(HeapTypeNameType_Spec.name);
+    if (PyObject_SetAttrString(HeapTypeNameType,
+                               "__qualname__", spec_name) < 0) {
```

Comment:
This is an interesting test.
Would it make sense to test the *original* qualname first, then assign a different one and test that `PyType_GetQualName` returns the replacement?

Reply:
Haha. I try to get some qual names in _testcapimodule. But I can get the qual name with prefix.
The reason is `res->ht_qualname = res->ht_name;` in https://github.com/python/cpython/blob/main/Objects/typeobject.c#L3439.
I don't know the reason why the qual name need to reomve the prefix too.
So I set a qual name with prefix to test this C API, somthing like a mock behavior~

Reply:
Right, C-API classes don't have distinct qualnames.
IMO, the best thing to do here would be:
- Test the *original* qualname (same as the short name)
- Change the qualname
- Then test the new qualname.

Ideally, the same thing would be tested for `PyType_GetName` as well – it's not that different from this function.

Reply:
> Right, C-API classes don't have distinct qualnames.
> IMO, the best thing to do here would be:
> 
> * Test the _original_ qualname (same as the short name)
> * Change the qualname
> * Then test the new qualname.

Make sense. How about this updated PR :)

> Ideally, the same thing would be tested for `PyType_GetName` as well – it's not that different from this function.

Got it. I will update enhance this test case later.